### PR TITLE
[1.21.3] Fix unbound tags issue in registry snapshot application

### PR DESCRIPTION
**Pull request by dhyces**
Pretty much part 2 of #1651, due to missed registry freeze issue that didn't crop up in the added test. Hopefully the only other location where registry freeze needs to be messed with to allow named holder sets in static registry objects.

### CODE REVIEW

**Comment by Shadows-of-Fire** [comment]
Can you try adding a test with a statically declared ArmorMaterial that uses a custom tag?  Trying to see if we can force the issue to be tested.

**Comment by dhyces** [comment]
Tried a static variable, still succeeded just fine in the test after removing the fix. I'll look into whether there's some way to replicate this later.

**Comment by dhyces** [comment]
> Can you try adding a test with a statically declared ArmorMaterial that uses a custom tag? Trying to see if we can force the issue to be tested.

Confirmed that this only seems to occur if a registry-related exception is thrown, causing the true error to be suppressed by the tag binding exception in freeze.

File: src/main/java/net/neoforged/neoforge/registries/GameData.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ public static void unfreezeData() {
     public static void freezeData() {
         LOGGER.debug(REGISTRIES, "Freezing registries");
         BuiltInRegistries.REGISTRY.stream().filter(r -> r instanceof MappedRegistry).forEach(r -> {
-            // HolderSet.Named may be used for registry objects, vanilla binds these tags so freeze won't throw for unbound tags
+            // HolderSet.Named may be used for registry objects, vanilla binds these tags so freeze doesn't throw for unbound tags
             ((MappedRegistry<?>) r).bindAllTagsToEmpty();
```
Comment (Shadows-of-Fire) [inline src/main/java/net/neoforged/neoforge/registries/GameData.java:70]: With the patch in `MappedRegistry` above, do we still need the `bindAllTagsToEmpty` call here?  Are they trying to achieve the same purpose?

Comment (dhyces) [reply src/main/java/net/neoforged/neoforge/registries/GameData.java:70]: Different purposes. Binding all tags to empty here allows tags to be used and they'll be properly rebound on world load. The clear call is used for when a registry error occurs, where it just removes all unbound holder sets, effectively reducing them to the same result as you would get from `HolderSet.emptyNamed`.

Comment (Shadows-of-Fire) [reply src/main/java/net/neoforged/neoforge/registries/GameData.java:70]: Makes sense, are we certain there are no other problematic call sites where we call `freezeRegistry` that might need special treatment?

Comment (dhyces) [reply src/main/java/net/neoforged/neoforge/registries/GameData.java:70]: I ran a search after this last error and didn't see any other problematic call sites, but I'll double check when I get back

Comment (dhyces) [reply src/main/java/net/neoforged/neoforge/registries/GameData.java:70]: Everything looks good, I'll resolve this and re-request review

**Review by Shadows-of-Fire — APPROVED** [review]
Hopefully this doesn't regress again in the future, unfortunate that the trigger is too difficult to exercise in testing.

**Pull request merged**